File: app/src/main/java/com/example/josegarcia/todaymeal/adapter/RecipeListAdapter.java
package com.example.josegarcia.todaymeal.adapter;

import android.graphics.drawable.BitmapDrawable;
import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.ListAdapter;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.josegarcia.todaymeal.R;
import com.example.josegarcia.todaymeal.helper.ImageCache;
import com.example.josegarcia.todaymeal.model.Recipe;
import com.example.josegarcia.todaymeal.views.SelectRecipe;
import com.squareup.picasso.Picasso;

public class RecipeListAdapter extends ListAdapter<Recipe, RecipeListAdapter.RecipeViewHolder> {
    private SelectRecipe onSelectRecipeListener;

    public RecipeListAdapter(SelectRecipe onSelectRecipeListener) {
        super(new RecipeDifferCallback());
        this.onSelectRecipeListener = onSelectRecipeListener;
    }

    @NonNull
    @Override
    public RecipeViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int type) {
        View view = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.recipe_card, viewGroup, false);
        return new RecipeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecipeViewHolder recipeViewHolder, int position) {
        Recipe recipe = getItem(position);
        recipeViewHolder.bind(recipe);
    }

    class RecipeViewHolder extends RecyclerView.ViewHolder {
        private final ImageView foodPicture = itemView.findViewById(R.id.food_image);
        private final TextView foodName = itemView.findViewById(R.id.recipe_name);

        RecipeViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        void bind(final Recipe recipe) {
            if (recipe.getImage().isEmpty()) {
                foodPicture.setImageResource(R.drawable.cooking_table);
            } else {
                Picasso.get()
                        .load(recipe.getImage())
                        .placeholder(R.drawable.loading_image_animation)
                        .into(foodPicture);
            }
            foodName.setText(recipe.getName());
            itemView.setOnClickListener(view -> {
                if (!recipe.getImage().isEmpty()) {
                    ImageCache.getInstance().setBitmap(((BitmapDrawable) foodPicture.getDrawable())
                            .getBitmap());
                }
                onSelectRecipeListener.onSelectRecipeListener(recipe);
            });
        }
    }
}

File: app/src/main/java/com/example/josegarcia/todaymeal/adapter/RecipeListAdapter.kt
package com.example.josegarcia.todaymeal.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.josegarcia.todaymeal.R
import com.example.josegarcia.todaymeal.model.Recipe
import com.example.josegarcia.todaymeal.views.RecipeListFragmentDirections
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_recipe.view.recipe_score as recipeScore
import kotlinx.android.synthetic.main.recipe_servings.view.recipe_servings as recipeServings
import java.util.*
import kotlinx.android.synthetic.main.item_recipe.view.recipe_image as recipeImage
import kotlinx.android.synthetic.main.item_recipe.view.recipe_name as recipeName

class RecipeListAdapter :
    ListAdapter<Recipe, RecipeListAdapter.RecipeViewHolder>(RecipeDifferCallback()) {

    override fun onCreateViewHolder(viewGroup: ViewGroup, type: Int): RecipeViewHolder {
        val view = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.item_recipe, viewGroup, false)
        return RecipeViewHolder(view)
    }

    override fun onBindViewHolder(recipeViewHolder: RecipeViewHolder, position: Int) =
        recipeViewHolder.bind(getItem(position))

    class RecipeViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {

        fun bind(recipe: Recipe) {
            loadImage(recipe)
            setClickListener(recipe)
            itemView.recipeName.text = recipe.name
            itemView.recipeScore.score = recipe.score
            itemView.recipeServings.text = String.format(Locale.US, "%d", recipe.servings)
        }

        private fun loadImage(recipe: Recipe) {
            if (recipe.image.isEmpty()) {
                itemView.recipeImage.setImageResource(R.drawable.cooking_table)
            } else {
                Picasso.get()
                    .load(recipe.image)
                    .into(itemView.recipeImage)
            }
        }

        private fun setClickListener(recipe: Recipe) {
            itemView.setOnClickListener {
                val dest =
                    RecipeListFragmentDirections.actionRecipeListFragmentToRecipeDescriptionFragment(
                        recipe,
                        recipe.name
                    )
                findNavController(itemView).navigate(dest)
            }
        }
    }
}
